### ISSUE TITLE
Feature Request: Allow OpenVPN server to override common name/username of a connection

### ISSUE THREAD

**Issue by jkroepke** [OpenVPN/openvpn#299]
## Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) 
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request

---

I'm creating this feature request based the discussion from

- https://github.com/OpenVPN/openvpn/issues/296#issuecomment-1485693101
- https://patchwork.openvpn.net/project/openvpn2/patch/20221010071229.7935-1-gert@greenie.muc.de/

---

I would like to share my use-case to give an better understand what I'm currently try to do:

OpenVPN 2.6.0+ give me the capabilities to delegate the authentication to external SSO/IDP by using the deferred auth mechanism. Currently, I'm develop a server-sided program with allows OpenVPN SSO Authentication with initially supported AzureAD and could also support Keycloak later. The auth flow based on the [OAuth 2.0 device authorization grant flow](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-device-code).

In that case from OpenVPN point of view, the connection from a user is initially credentials-less/identity-less. (it up the admin later to use an additional client-certificate). Client credentials (username/password) are not entered at OpenVPN anymore, instead the server send a browser based login page to the client using the `OPENURL` feature..On the login page, the user enters username, password and MFA. Since the login is inside the browser, additional protection like conditional access can be applied (depends on the feature set of the IDP).

On server side, the `auth-user-pass-verify` script continues pulling the IDP endpoint and wait until the user is logged-in or the timeout is reached.

Path to the `auth-user-pass-verify`: https://github.com/jkroepke/openvpn-auth-azure-ad/tree/v2. The v2 is not released yet. v1 is using the OpenVPN management interface together with abusing the dynamic challenge method which I would like to deprecate.

After reading the discussion to `[Openvpn-devel] TLS: do not lock empty usernames`, I though If changing the username is a no-go in after a successful initial authentication, what about changing the username on a initial authentication and push the information back the user? In my case, the `auth-user-pass-verify` script gets the information which user has logged-in. If the script cloud override the username on the connection, the idea could benefit together with the `username-as-common-name` option from a lot of existing OpenVPN features like `ccd`, `auth-token` or connection statistics and there is a 1:1 relation between the authenticated user and the logs on the OpenVPN server.

It should resolve the caveats from the [discussion](https://patchwork.openvpn.net/project/openvpn2/patch/20221010071229.7935-1-gert@greenie.muc.de/), since the username is changed on an authentication level and can be locked as currently designed.

**Comment by jkroepke** [comment]
Please let me know, If I should repost this on the openvpn-devel mailing list. I'm why more familiar with GitHub.

**Comment by jkroepke** [comment]
> auth-token-user exist but only a subset of clients versions actually supports it.

At least for my use-case, this isn't a problem. I have also a dependency against SSO OPENURL, which also only a subset of clients understand. 

---

I understand that `auth-token` was introduced for MFA authentication. But since users can configure deferred auth and SSO/OPENURL without writing a management client, `auth-token` could also use a "remember me" functionality to skip the whole SSO login process on TLS soft reset (`reneg-sec`).

Something like `set-auth-username` looks good to read, but I'm far away to understand the consequences here.

**Comment by jkroepke** [comment]
Another use-case for override clients username would be statistics.

If client certificates are omit (because of WebAuth), then OpenVPN status looks like:

```
root@localhost:~# cat  /var/run/openvpn-server/status-tcp8001.log
OpenVPN CLIENT LIST
Updated,2023-12-27 08:10:21
Common Name,Real Address,Bytes Received,Bytes Sent,Connected Since
UNDEF,219.93.0.130:48881,1784,3808,2023-12-27 08:10:02
UNDEF,219.93.0.130:58579,1558,3660,2023-12-27 08:10:16
ROUTING TABLE
Virtual Address,Common Name,Real Address,Last Ref
10.8.0.36,UNDEF,219.93.0.130:58579,2023-12-27 08:10:19
10.8.0.37,UNDEF,219.93.0.130:48881,2023-12-27 08:10:20
GLOBAL STATS
Max bcast/mcast queue length,2
END
```

Since we do not specify a username anywhere, which can later be used as CN, it is not clear which user is which client (Common Name = UNDEF)

**Comment by andreas-p** [comment]
@schwabe You're mentioning CHANGING the username, but apparently it is still unset when the deferred auth takes place. So to keep the the "bag of other issues" as small as possible, checking for previously unset username might help.
Deferred auth isn't really complete if the user is only authenticated as "some unnamed dude we know".

**Comment by schwabe** [comment]
@andreas-p internally that is still changing from the empty username `""` to some other username. The username gets transmitted very early in the OpenVPN handshake and OpenVPN currently sticks that username whether it is empty or something else. Allowing to change from `""` to another username still needs to check if that is well handled and not same parts of OpenVPN still use `""` or (printed as UNDEF but really `""`) as username.

**Comment by jkroepke** [comment]
Could be a display name field a possible solution? An additional, server exclusive field used in log output and stats?

**Comment by aprudnev** [comment]
Actually, it is important. We just created POC of WebAuth and OpenVPN  (using Ping Federate as IdP). It works well. But when we try to match IP assigned to the session, and user's name, we have a problem, and it is serious. User's name is reported by plugin (openvpn-auth-oauth2). IP is reported by openvpn. There is not a single field which allows to match. And it is absolutely critical to know user name in every openvpn log message.

it can be any name, better sub (from the token), but it is not critical - critical is to always have some name from the OIDC token in OpenVPN expecially in the messages related to IP assignments. As it is a very common case - track IP / time / port to the end user who used it (we have such trackings every few month, for example).

**Comment by aprudnev** [comment]
Other interesting problem related to logout (as usual, login/connect is way easier vs logout/disconnect). We want to clear session on explicit disconnect - if user disconnected he DISCONNECTED. ideally I would like to send user to logout page of our / his IdP. I do not see how to do it.

Else user stay connected even if disconnected - if someone takes his device and click on VPN he will be connected even if his MFA device is not there. Which violates security. At least, user need a choice to just 'disconnect temporarily' and 'logout from SSO session'.

**Comment by schwabe** [comment]
@aprudnev That is more a problem of your authentication stack than OpenVPN itself. For what you are describing you probably want to use an authentication backend that generates the auth-token itself and tracks/logs it and can invalidate it rathre than to rely on auth-gen-token which does not have that invalidation support.

**Comment by aprudnev** [comment]
This is partially true. Our POC uses openvpn-auth-oauth2 plugin. It all works pretty well, but we (still, even as I added 'use user name as common name' option) still do not see exact user name in the logs (system still report me name from the certificate), and it is a huge security issue. If someone uses certs / 1 cert per client, it makes perfect sense, but it is actually OIDC token which provides proof of authentication.

So for invalidation, the only issue we see is that we would like to have a way to call some URL on disconnect, exactly as we called other URL on connect for authentication. We expected this problem, and It is possible that it have easy solution, it is too early to consider anything (as our POC went alive just today). But this is a very common problem with all SSO / OIDC systems, so it is possible that some support in the server may help here, I just do not know yet. As I said, we just started to work on this, and we did not find / use all possible options. Both problems (user names in the logs; session disconnects as full disconnect) was expected and well predicted, as they are very common for SSO systems.

**Comment by schwabe** [comment]
I think the openvpn-auth-oauth2 plugin should just log username and so on for that connection. Yes setting the OpeNVPN username would be nice in this scenario but that does not excuse the plugin from not properly logging all the required information.

**Comment by aprudnev** [comment]
It logs username. BUT, it is absolutely critical to link IP assigned by
OPENVPN (and session) with user name. One of security requirements is _all
access events must be logged, together with - external IP, user-s name.
internal IP, in and out time_.  IT is necessary to investigate any
security event / violation / etc. We managed to do it for OpenVPN @ PFSense
(except, it do not send client-s IP by Radius and it actually is a big
problem), and here , in test system which runs OVPN 2.6.8 +
openvpn-auth-oauth2 with PingID/PingFederate as IdP. we see user names in
one place, and IP and session, in another, without an easy way to link
these events. OVPN reports user name as certificate CN or empty (if I add
an option to use username as common name). And it makes security violated
-  no way to find who had IP 10.25.25.106 today at 18:00 PST (for example).

**Comment by schwabe** [comment]
@aprudnev OpenVPN provides that information in its logs and also to plugins.

**Comment by aprudnev** [comment]
Can OpenVPN use mapped attributes from OIDC as user name in it's logs? Of
course, if it provides IP and session start/end to the plugin, plugin can
log it too (and plugin can, maybe better of all, make logouts and/or
invalidate sessions if they are idle for some time), but OpenVPN
already try to log client-s name (I see that it use CN from certificate, or
use user name if it runs user-s authentication thru the Radius), so it is
very logical to use sub or username from SSO token for the same purpose.

I was able to get username in plugin logs; get IP-s in openvpn logs; no way
to link them together yet (at least I did not see how can I do it).

**Comment by jkroepke** [comment]
Hi all,

first of all, I hope we can discuss this feature request more on a technical level rather than on a emotional level.

Discussion around the `auth-token` is more a thing, that should be solved inside `openvpn-auth-oauth2`. I had a recent [discussion](https://github.com/OpenVPN/openvpn/issues/506) here, but I like more the current idea, that auth-token acts more as an Kerberos ticket. And sorry here for the overload, but it looks like I have to document more precisely which feature from OpenVPN can be used in which scenario. For any future discussion points around re-authentication, please create a issue `openvpn-auth-oauth2`.

I would like to go a step to the initial root cause, that management clients/plugins are unable to set an username/common name for a connection.

I fell that the username/common name is missing in the OpenVPN logs is not a deal breaker. And I would agree with you @schwabe, having the username in the OpenVPN is more "nice-to-have". I can add the client IP to the logs as well, since the ClientIP+Port can be used as ID between both solutions.

This real issue is, without an username:

- `ifconfig-pool-persist` not longer work
- `client-config-dir` not longer work - can be mitigated via `openvpn-auth-oauth2`
- The content of `status` doesn't help at all.
   ```
   OpenVPN CLIENT LIST
   Updated,2024-02-27 10:27:47
   Common Name,Real Address,Bytes Received,Bytes Sent,Connected Since
   UNDEF,192.168.65.1:37239,4318,7750,2024-02-27 10:24:44
   UNDEF,192.168.65.1:49624,1296,2394,2024-02-27 10:27:23
   ROUTING TABLE
   Virtual Address,Common Name,Real Address,Last Ref
   100.64.0.2,UNDEF,192.168.65.1:37239,2024-02-27 10:24:46
   100.64.0.3,UNDEF,192.168.65.1:49624,2024-02-27 10:27:25
   GLOBAL STATS
   Max bcast/mcast queue length,0
   END
   ```
- Gather any statistics through Management Interface does not provide useful information.

> That is more a problem of your authentication stack than OpenVPN itself. 

You are correct. It's an problem the authentication stack. But the problem can be resolved if OpenVPN server would provide a capability to set an Display Name / Username on a connection. If the Feature Request is implement, the problem is resolved, too.

OpenVPN fully supports [WEBAUTH](https://github.com/OpenVPN/openvpn3/blob/master/doc/webauth.md). It offers a modern authentication and is a good replacement for PKI based authentication. The SSO provider can offer any modern authentication as well. For example: FIDO2 Token, WebAuth, MFA via Push or biometric (e.g. Fingerprint). The SSO Provider can decide different action. For example: Deny logins outside of Germany or deny login on company unmanaged devices. 

Manage a separate PKI for OpenVPN would be obsolete.

---

The standalone WEBAUTH authentication system currently lacks the comprehensive features offered by VPN configurations utilizing client certificates. I'm interested in exploring how WEBAUTH can be enhanced to incorporate all the functionalities available in OpenVPN, ensuring that users can leverage the full spectrum of features even when exclusively using WEBAUTH for authentication.

**Comment by schwabe** [comment]
So your real issues are basically a problem for your setup since you still rely a lot on OpenVPN's internal management features and having them work with these features, we probably need to implement a feature to set a username.

The reason that this hasn't been really done yet is that our own implementation that heavily uses webauth and external auth system also does the same heavy lifting for the other features. Instead of relying on ifconfig-pool-persist and client-config-dir, these client specific settings are also provided by the backend via the management interface. So we only use the auth-token as assigning the client a session id and basically do everything else in the backend.

So in summary, yes this is a valid feature request. But currently the lack of this feature has little consequence for most people using very advanced authentication stacks as they are not depending on the internal features that require support for that. So it currently does not get the attention that it might need. As in most OSS projects features request etc get attention priority if a developer/submitter is interested in them (either personal or because their company is interested in that feature).

**Comment by aprudnev** [comment]
Good discussion, thanks.

One more. maybe last,  notice. It can be very useful if openjvpn can use
groups or roles, provided by plugins like oauth2 one, to decide on IP
setup. For example, we may want to assign  different IP pools to the
regular users and/or DevOps users. Today we do it by - configure few OVPN
instances, configure few RADIUS instances, have different groups allowing
access, and so on. Which is a lot of headache. This can be (maybe) done by
using certificate based authentication, but it is not easy to combine with
MFA which is , today, security requirement. So, if plugins can provide not
just username/displayname, but something which can be used for
configuration, it can be even better.

And yes, using WEBAUTH opens a huge set of options not available by
traditional authentications. So it is a very big step forward.

**Comment by schwabe** [comment]
@aprudnev what you are describing is way beyond what openvpn internally does. But this can implemented with an external authentication like I described. In fact, the commercial OpenVPN Access Server implements all these features with OpenVPN but managing all this things outside of OpenVPN itself. And it is unlikely that these features make it into OpenVPN itself since people either have a full custom solution like Access Server or a small setup that does not require this.

**Comment by aprudnev** [comment]
AS supports saml only, so they can be interested in idc/oauth integration
into AS version, too.  (SAML is actually terrible protocol, very annoying
to configure and support and not easy to scale).

As for us, we use ovpn mostly with pfsense virtual appliances (which are de
facto freebsd). We run AS in past but dropped it because of high price and,
more important, absence of many features like 'clients ip in radius rq',
'exposure of https port to Internet', 'single service per system' (we run 2
- 4 ovpn services per appliance, and no one is exposed to internet
scanners). This OIDC POC is pilot project to see, how can we move from
radius to webauth, and just first results are very positive, but this all
is not for most companies who better pay for the licenses and professional
support. I do not expect conflict of interests here, vice versa.

**Comment by schwabe** [comment]
@jkroepke sure external patches are accepted. I am just pointing out that almost everyone who developed for OpenVPN was part of a company and even other companies that would integrete OpenVPN in their products like Astaro in the past would choose the same approach like Access Server would use. So the effort in bringing this into OpenVPN itself from developers just hasn't been there. If someone would pick up that topic and provide patches then we would pick that up.

**Comment by aprudnev** [comment]
Having more functions outside of core code is, maybe, good idea, as we see
the same in K8S, for example. Problem is that information must be available
between components. If core assigns IP or makes a decision to do it, it
needs all information from the plugin (all attributes of the connection).
If a plugin or script does it, it needs all information from the core. Just
now, information is separated - plugin have attributes but do not know,
what core did (assigned IP, called script, and so on), core has not
information from plugin. Logs are just one example of 'collateral damage'
because of this.

Let's say, we want to assign VLAN and IP based on user's attributes (roles,
groups, similar) - a very common and reasonable task, actually. It does not
matter which exact component does it - it matters that this component needs
_all available information_. And just now, no one such component exists.

**Comment by schwabe** [comment]
@aprudnev what you want exists, solution like Access Server or other commercial solution have all that information.

**Comment by jkroepke** [comment]
I would appreciate that any future discussion, which are not in scope on the initial post are continue in a separate issue..

@aprudnev since you are prefer mail anyways, you can also use the mailing list: https://openvpn.net/community-resources/mailing-lists/

thanks!

**Comment by choseh** [comment]
Hi everyone, we're having the same requirement here and would appreciate an option to override the username. We're currently on an ldap setup that assigns users on ipsets based on their group memberships but want to move to Entra-I/Azure auth. Cuirrently there is no safe way to pass the username to the second "get-group-membership-and-assign-ipset" script. Only option would be to run multiple instances of the oauth2 script which is not a great option. Please reconsider adding an option.

**Comment by schwabe** [comment]
Here is a patch on gerrit that allows overriden the internal username of a connection that will also be used by auth-gen-token. This allows changing the username as part of management interface and client-connect script. Technically also as part of client-config-dir but that seems a weird use case: https://gerrit.openvpn.net/c/openvpn/+/872

Please check if that addresses your use cases.

**Comment by jkroepke** [comment]
Hi @schwabe,

I'm not familiar with gerrit. Does gerrit offer a git url to an branch where is patches are applied? Or do I have to manually apply the change by using a patch file?

I would like to test it out and give feedback somewhere, but I do not have an gerrit account.

**Comment by flichtenheld** [comment]
@jkroepke Clicking on "Download" (above the file list) should give you a list of options how to check out/apply the patch. This is all possible with anonymous access, no account required.

**Comment by aprudnev** [comment]
I will experiment. We do use this plugin with OpenVPN (with great success).
Through, we can link user-s login event with user-s login name using
client-s IP + port (so we have necessary statistics) but it's always better
tohave them in the single record in the logs.

**Comment by jkroepke** [comment]
# Feedback

## Docs

It takes some time to understand how it works. It was not clear to me that `override-username` accept a parameters. It was not clear to, because I expect a pattern from the OpenVPN man page. Looking to other options, e.g. `auth-user-pass-verify cmd method`, it clearly indicates that `auth-user-pass-verify` accept 2 parameters. 

## Questions

Are there any limitations aside from length? Is UTF-8 supported for username?

## What is working

Running `status 2` print the username from the `override-username` command. Thats great and huge benefit for setup with credentials less clients.

## Compatibility with username-as-common-name

There is a compatibility issue with `username-as-common-name` in combination with `client-config-dir`. I guess the username change appears after OpenVPN looks for an file in `client-config-dir`. People may expect, if `override-username username` is set, OpenVPN looks for a file named `/etc/openvpn/ccd/username`. 



## client-connect

The value of `override-username` is not available as environment variable in client-connect script. Looks like the same issue as client-config-dir.

## TLS Auth Error on session renegotiation

To test all capabilities, I lower the value of `reneg-sec`. I observe `2025-01-20 20:04:44 172.18.0.1:63189 MULTI: Note, override-username changes username from '' to 'mail@jkroepke.de'` in the server log, but the re auth fails. `2025-01-20 20:05:40 172.18.0.1:63189 TLS: Username/Password authentication deferred for username 'mail@jkroepke.de' [CN SET]` and `
2025-01-20 20:05:40 172.18.0.1:63189 TLS Auth Error: TLS object CN attempted to change from '' to 'mail@jkroepke.de' -- tunnel disabled`

Here are my server logs

<details>
<summary>Details</summary>


```
2025-01-20 20:04:34 Initialization Sequence Completed
2025-01-20 20:04:42 172.18.0.1:63189 peer info: IV_VER=3.8.2connect3
2025-01-20 20:04:42 172.18.0.1:63189 peer info: IV_PLAT=mac
2025-01-20 20:04:42 172.18.0.1:63189 peer info: IV_NCP=2
2025-01-20 20:04:42 172.18.0.1:63189 peer info: IV_TCPNL=1
2025-01-20 20:04:42 172.18.0.1:63189 peer info: IV_PROTO=990
2025-01-20 20:04:42 172.18.0.1:63189 peer info: IV_MTU=1600
2025-01-20 20:04:42 172.18.0.1:63189 peer info: IV_CIPHERS=AES-128-CBC:AES-192-CBC:AES-256-CBC:AES-128-GCM:AES-192-GCM:AES-256-GCM:CHACHA20-POLY1305
2025-01-20 20:04:42 172.18.0.1:63189 peer info: IV_AUTO_SESS=1
2025-01-20 20:04:42 172.18.0.1:63189 peer info: IV_GUI_VER=OCmacOS_3.4.9-4830
2025-01-20 20:04:42 172.18.0.1:63189 peer info: IV_SSO=webauth,crtext
2025-01-20 20:04:42 172.18.0.1:63189 peer info: IV_BS64DL=1
2025-01-20 20:04:42 172.18.0.1:63189 TLS: Username/Password authentication deferred for username '' [CN SET]
2025-01-20 20:04:42 172.18.0.1:63189 TLS: move_session: dest=TM_ACTIVE src=TM_INITIAL reinit_src=1
2025-01-20 20:04:42 172.18.0.1:63189 TLS: tls_multi_process: initial untrusted session promoted to semi-trusted
2025-01-20 20:04:42 MANAGEMENT: CMD 'client-pending-auth 0 1 "WEB_AUTH::https://joe-nb-2:9000/oauth2/start?state=e" 180'
2025-01-20 20:04:42 SENT CONTROL []: 'AUTH_PENDING,timeout 29' (status=1)
2025-01-20 20:04:42 SENT CONTROL []: 'INFO_PRE,WEB_AUTH::https://joe-nb-2:9000/oauth2/start?state=e' (status=1)
2025-01-20 20:04:42 172.18.0.1:63189 Control Channel: TLSv1.3, cipher TLSv1.3 TLS_AES_256_GCM_SHA384, peer temporary key: 253 bits X25519
2025-01-20 20:04:42 172.18.0.1:63189 [] Peer Connection Initiated with [AF_INET]172.18.0.1:63189
2025-01-20 20:04:42 172.18.0.1:63189 PUSH: Received control message: 'PUSH_REQUEST'
2025-01-20 20:04:43 172.18.0.1:63189 PUSH: Received control message: 'PUSH_REQUEST'
2025-01-20 20:04:43 MANAGEMENT: CMD 'client-auth 0 1'
2025-01-20 20:04:44 172.18.0.1:63189 MULTI_sva: pool returned IPv4=100.64.0.2, IPv6=(Not enabled)
2025-01-20 20:04:44 172.18.0.1:63189 MULTI: Note, override-username changes username from '' to 'mail@jkroepke.de'
2025-01-20 20:04:44 172.18.0.1:63189 MULTI: Learn: 100.64.0.2 -> 172.18.0.1:63189
2025-01-20 20:04:44 172.18.0.1:63189 MULTI: primary virtual IP for 172.18.0.1:63189: 100.64.0.2
2025-01-20 20:04:44 172.18.0.1:63189 SENT CONTROL []: 'PUSH_REPLY,route-gateway 100.64.0.1,topology subnet,ping 10,ping-restart 60,auth-token-user bWFpbEBqa3JvZXBrZS5kZQ==,ifconfig 100.64.0.2 255.255.255.0,peer-id 0,auth-tokenSESS_ID,auth-token-user bWFpbEBqa3JvZXBrZS5kZQ==,cipher AES-256-GCM,protocol-flags cc-exit tls-ekm dyn-tls-crypt,tun-mtu 1500' (status=1)
2025-01-20 20:04:45 172.18.0.1:63189 Data Channel: cipher 'AES-256-GCM', peer-id: 0
2025-01-20 20:04:45 172.18.0.1:63189 Timers: ping 10, ping-restart 120
2025-01-20 20:04:45 172.18.0.1:63189 Protocol options: explicit-exit-notify 1, protocol-flags cc-exit tls-ekm dyn-tls-crypt
2025-01-20 20:05:40 172.18.0.1:63189 TLS: soft reset sec=58/58 bytes=556/-1 pkts=12/0 aead_limit_send=25/60129542137 aead_limit_recv=12/60129542137
2025-01-20 20:05:40 172.18.0.1:63189 peer info: IV_VER=3.8.2connect3
2025-01-20 20:05:40 172.18.0.1:63189 peer info: IV_PLAT=mac
2025-01-20 20:05:40 172.18.0.1:63189 peer info: IV_NCP=2
2025-01-20 20:05:40 172.18.0.1:63189 peer info: IV_TCPNL=1
2025-01-20 20:05:40 172.18.0.1:63189 peer info: IV_PROTO=990
2025-01-20 20:05:40 172.18.0.1:63189 peer info: IV_MTU=1600
2025-01-20 20:05:40 172.18.0.1:63189 peer info: IV_CIPHERS=AES-128-CBC:AES-192-CBC:AES-256-CBC:AES-128-GCM:AES-192-GCM:AES-256-GCM:CHACHA20-POLY1305
2025-01-20 20:05:40 172.18.0.1:63189 peer info: IV_AUTO_SESS=1
2025-01-20 20:05:40 172.18.0.1:63189 peer info: IV_GUI_VER=OCmacOS_3.4.9-4830
2025-01-20 20:05:40 172.18.0.1:63189 peer info: IV_SSO=webauth,crtext
2025-01-20 20:05:40 172.18.0.1:63189 TLS: Username/Password authentication deferred for username 'mail@jkroepke.de' [CN SET]
2025-01-20 20:05:40 172.18.0.1:63189 TLS Auth Error: TLS object CN attempted to change from '' to 'mail@jkroepke.de' -- tunnel disabled
2025-01-20 20:05:40 172.18.0.1:63189 Delayed exit in 5 seconds
2025-01-20 20:05:40 172.18.0.1:63189 SENT CONTROL [UNDEF]: 'AUTH_FAILED' (status=1)
2025-01-20 20:05:40 172.18.0.1:63189 SENT CONTROL []: 'AUTH_FAILED' (status=1)
2025-01-20 20:05:40 172.18.0.1:63189 Control Channel: TLSv1.3, cipher TLSv1.3 TLS_AES_256_GCM_SHA384, peer temporary key: 253 bits X25519
2025-01-20 20:05:41 MANAGEMENT: CMD 'client-auth 0 3'
2025-01-20 20:05:45 172.18.0.1:63189 SIGTERM[soft,delayed-exit] received, client-instance exiting
```



</details>

Server Config: https://github.com/jkroepke/openvpn-auth-oauth2/blob/ede97912e4a6593235e340f0c6b3b00e569c7957/tests/docker-compose.yaml#L53-L98
Client Config: https://github.com/jkroepke/openvpn-auth-oauth2/blob/ede97912e4a6593235e340f0c6b3b00e569c7957/tests/docker-compose.yaml#L103-L121

**Comment by schwabe** [comment]
@jkroepke the man page changes of the commit strongly advises against using username-as-common-name with this option and also points out that the option is only done after these options. Having `client-config-dir` use the name of the overridden username would complicate things a lot as you would need to have a two step parsing. And also management and client-connect can defer, which complicates things. I also do not see the use-case to be honest. You need client-connect or management realistically to set this option anyway. And if you use client-connect or management to set one option, you can just set all options and don't need client-config-dir.

As for the tls auth errors on renegotiation, I am pretty sure that this is a side effect of using `username-as-common-name`. I might just update the patch to make the option incompatible with `username-as-common-name`. Using them both together feels like just asking for a lot of weird corner cases, especially considering all the deferred options.

As for base64 and so on, the option will just take whatever you put in there, so using utf-8 should work fine.

**Comment by jkroepke** [comment]
I can confirm, that remove `username-as-common-name` resolves my issue. I tested with OpenVPN client 3.6.0 and 2.6.12

**Comment by jkroepke** [comment]
> I am overlooking. I am just looking at this mostly from the "management control alsmost everything approach"

The only use case I'm aware of it is that something is not support by the management control software and sysadmin may use client-config-dir to sideload additional configuration for connections. 

It's fair to not officially support such unhealthy situations. The documentation clears explain what is possible with `override-username`. 

The main paint point was that systems can't identity connection from openvpn server log point of view. This is now possible, since the username change this logged and visible on the status monitor as well.

**Comment by jkroepke** [comment]
Additionally, connections with empty usernames are supported as will. I notice that the session state is now `Authenticated` instead `AuthenticatedEmptyUser` if override-username is in use. 

It seems like `override-username` is a solution for https://github.com/OpenVPN/openvpn/issues/296 as well.

**Comment by aprudnev** [comment]
And I do not want access server, it is way less flexible and is too
expensive and unsafe (as too many hackers try to broke it). We run 99% on
open source and it show its strength all these years (same for log
processing, same for monitoring). We run few AS in the past, its better vs
anyconnect but way behind community version in many ways, and no OIDC +
subscription license and open port 443 are major factors.

**Comment by schwabe** [comment]
@jkroepke the comment was mainly meant that I looking at this problem from the perspective that I am using. So if there is still a problem with another use case, someone else might need to explain me the use case so I can properly understand it.

**Comment by schwabe** [comment]
@aprudnev your comment about Access Server being unsafe as too many hackers broke it without any evidence is just bad mouthing. Please do not such statement without any proof.

**Comment by aprudnev** [comment]
AS problem is that it expose port 443 with an easy detection that 'AS
server here', and it always creates honeypot for the hackers. AS itself is
excellent in terms of security, ease to use, features. But its just this
exposure - if I scan IP I can easily find AS servers around. We used it for
5 years, and it attracted hackers (and security scanners) 24x7.

In addition, it do not support OIDC yet (I can be wrong) only SAML. Which
is way more complicated and cause a lot of headache (manual cert updates,
for example).

What we created instead - you can not scan it by IP, you can not detect
that it is openvpn, it uses OIDC SSO to deliver profiles and to connect to
VPNs. So our IDS systems do not see anyone scanning it or trying to break
in.

But, in terms of security and ease if use, I always recommend AS for non IT
companies who need commercial solutions.

**Comment by jkroepke** [comment]
> [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/1560587?u=32b3431ddec0a7f93ba321721812f74ccb48fa21&amp;v=4&amp;size=80">@jkroepke](https://github.com/jkroepke) I updated the version on Gerrit to also play nicely with `--username-as-common-name`. I still don't recommend that combination but it should work now. Can you try again?

It is working now.

I also observe that the overridden username value is used in context of `ifconfig-pool-persist`.  This is at least a work around for users to assign static IP from the IP pool, if the management client does not support that feature and client-config-dir is not available in that context. If this is intended, it might be worth to document.

```
openvpn-1  | 2025-01-25 10:19:10 172.18.0.1:64330 MULTI: Note, override-username changes username from '' to 'github_jkroepke.de'
openvpn-1  | 2025-01-25 10:19:10 172.18.0.1:64330 MULTI: Learn: 100.64.0.2 -> github_jkroepke.de/172.18.0.1:64330
openvpn-1  | 2025-01-25 10:19:10 172.18.0.1:64330 MULTI: primary virtual IP for github_jkroepke.de/172.18.0.1:64330: 100.64.0.2
```

Soft-reset works fine as well. Note sure, if the message `Replacing client provided username` is mandatory, if both values are equal. Theoretically, this messages on a re-auth is misleading, since `github_jkroepke.de` was not initially provided by client.

```
openvpn-1  | 2025-01-25 10:20:08 172.18.0.1:64330 TLS: Replacing client provided username 'github_jkroepke.de' with username from override-user 'github_jkroepke.de'
openvpn-1  | 2025-01-25 10:20:08 172.18.0.1:64330 TLS: Username/Password authentication deferred for username 'github_jkroepke.de' [CN SET]
openvpn-1  | 2025-01-25 10:20:08 172.18.0.1:64330 Control Channel: TLSv1.3, cipher TLSv1.3 TLS_AES_256_GCM_SHA384, peer temporary key: 253 bits X25519
openvpn-1  | 2025-01-25 10:20:09 MANAGEMENT: CMD 'client-auth 0 3'
openvpn-1  | 2025-01-25 10:20:23 172.18.0.1:64330 SENT CONTROL [github_jkroepke.de]: 'PUSH_REPLY,auth-tokenSESS_ID' (status=1)
```

**Comment by schwabe** [comment]
I will fix the rename from same user message. As for persist pool, it depends what username you are getting there on the order of things that are executed. Deferred vs non-deferred script and so on. While it probably writes with the overridden name all the time, I don't think it will always lookup the overridden name but I haven't deep dived into this.

**Comment by jkroepke** [comment]
hm, does it look like a timing thing?

Taking a look at my first post https://github.com/OpenVPN/openvpn/issues/299#issuecomment-2603207883 , it looks like 

Patchset 2
```
2025-01-20 20:04:44 172.18.0.1:63189 MULTI: Note, override-username changes username from '' to 'github_jkroepke.de'
2025-01-20 20:04:44 172.18.0.1:63189 MULTI: Learn: 100.64.0.2 -> 172.18.0.1:63189
2025-01-20 20:04:44 172.18.0.1:63189 MULTI: primary virtual IP for 172.18.0.1:63189: 100.64.0.2
```

e.g. no references to the overridden username at the IP assign logs.

VS

Patchset 4 + `username-as-common-name`
```
2025-01-25 10:19:10 172.18.0.1:64330 MULTI: Note, override-username changes username from '' to 'github_jkroepke.de'
2025-01-25 10:19:10 172.18.0.1:64330 MULTI: Learn: 100.64.0.2 -> github_jkroepke.de/172.18.0.1:64330
2025-01-25 10:19:10 172.18.0.1:64330 MULTI: primary virtual IP for github_jkroepke.de/172.18.0.1:64330: 100.64.0.2
```

Test environment changes: https://github.com/jkroepke/openvpn-auth-oauth2/commit/7507a8f3aac8578d3cd8b46424e8d7667ccef9b8

**Comment by cron2** [comment]
The last revision of the patch is now in master.  Please re-test and complain to me if I overlooked something :-)

**Comment by jkroepke** [comment]
Hi,

from my point of view, override username works fine. I also test `username-as-common-name`  even it's not recommend. I would like to let you know, if the client does not provide a common name, a username and `username-as-common-name` is set, then the common name is overwritten as well. The status command on the management interface reports this well.

Not sure, if this case is expected, since the behavior does not match the documentation. 

> Also ``--username-as-common-name`` will use the client provided username
  as common-name. It is recommended to avoid the use of the
  ``--override-username`` option if the option ``--username-as-common-name``
  is being used.

But it would be great, if that behavior could be forced in any case.

**Comment by schwabe** [comment]
Hm where do see that this is different from what is documented?

```
      --username-as-common-name
              Use the authenticated username as the common-name, rather than
              the common-name from the client certificate.
```

That says it will overwrite the commonname. And if you have an empty username then you end up with an empty common name as well. So what do you think should be changed or enforced here?

**Comment by aprudnev** [comment]
I do not see a problem here, You have per client certificate - then you use
name from certificate - or you have one certificate for all users + use
authentication. It is unlikely you use both.

But if in doubts just specify format like %C - common name, %A -
authenticated name; then allow to use formats like %A:%C or %A?%C (if %A
empty use %C)

**Comment by jkroepke** [comment]
@schwabe I was confused, because if username and common name is both empty, the value from `override-username` is not only set as username, it sets common name as will.

**Comment by schwabe** [comment]
Okay. Anyway. I will close this issue now since it has been implemented and improvement and bug to the feature are better handled in sperate tickets as this issue is becoming more and more convuluted.
